### PR TITLE
FDB backend support

### PR DESCRIPTION
**Pull request by ViramShah91**
UT:

FDB-1
FDB-2
FDB-3
FDB-4

### CODE REVIEW

File: models/yang/acme-system.yang [hunk 1]
```diff
@@ -0,0 +1,50 @@
+     module acme-system {
```
Comment (jeffyin-dell) [inline models/yang/acme-system.yang]: Why is this being added?

File: models/yang/Makefile [hunk 2]
```diff
@@ -1,21 +1,9 @@
-################################################################################
```
Comment (jeffyin-dell) [inline models/yang/Makefile]: Should probably leave this in...

File: models/yang/annotations/openapi_annotations/acme-system-annot.yang [hunk 3]
```diff
@@ -0,0 +1,17 @@
+module acme-system-annot {
```
Comment (jeffyin-dell) [inline models/yang/annotations/openapi_annotations/acme-system-annot.yang]: Why is this file being added?

File: models/yang/common/openconfig-aft-mpls.yang [hunk 4]
```diff
@@ -0,0 +1,117 @@
+submodule openconfig-aft-mpls {
```
Comment (jeffyin-dell) [inline models/yang/common/openconfig-aft-mpls.yang:1]: Are we supporting this?

**Review by jeffyin-dell — CHANGES_REQUESTED** [review]
I think you have a lot of extraneous files being added in this PR. Please re-review and update the PR.

Venkat > Yes, lot of YANG files are required for network-instance YANG compilation.

File: models/yang/Makefile (hunk 2 above)
Comment (ViramShah91) [reply models/yang/Makefile]: Done

File: models/yang/acme-system.yang (hunk 1 above)
Comment (ViramShah91) [reply models/yang/acme-system.yang]: I have deleted the file from the branch as it's not needed.

File: models/yang/common/openconfig-aft-mpls.yang (hunk 4 above)
Comment (ViramShah91) [reply models/yang/common/openconfig-aft-mpls.yang:1]: openconfig-aft is required for network-instance-yang.

File: models/yang/annotations/openapi_annotations/acme-system-annot.yang (hunk 3 above)
Comment (ViramShah91) [reply models/yang/annotations/openapi_annotations/acme-system-annot.yang]: I have deleted the file from the branch as it's not needed.

**Comment by ViramShah91** [comment]
> I think you have a lot of extraneous files being added in this PR. Please re-review and update the PR.

Done

File: src/translib/transformer/xlate_from_db.go [hunk 5]
```diff
@@ -150,7 +150,7 @@ func DbToYangType(yngTerminalNdDtType yang.TypeKind, fldXpath string, dbFldVal s
 		if res, err = strconv.ParseBool(dbFldVal); err != nil {
 			log.Warningf("Non Bool type for yang leaf-list item %v", dbFldVal)
 		}
-        case yang.Ybinary, yang.Ydecimal64, yang.Yenum, yang.Yidentityref, yang.Yint64, yang.Yuint64,     yang.Ystring, yang.Yunion:
+        case yang.Ybinary, yang.Ydecimal64, yang.Yenum, yang.Yidentityref, yang.Yint64, yang.Yuint64, yang.Ystring, yang.Yunion, yang.Yleafref:
```
Comment (venkatmahalingam) [inline src/translib/transformer/xlate_from_db.go]: This change should not be part of the PR, ask infra team to commit this change.

File: src/translib/transformer/xfmr_fdb.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	strconv.FormatInt(int64(ocbinds.OpenconfigNetworkInstance_NetworkInstances_NetworkInstance_Fdb_MacTable_Entries_Entry_State_EntryType_DYNAMIC), 10): SONIC_ENTRY_TYPE_DYNAMIC,
+}
+
+
+var YangToDb_fdb_tbl_key_xfmr KeyXfmrYangToDb = func(inParams XfmrParams) (string, error) {
+    var entry_key string
+    var err error
+
+    pathInfo := NewPathInfo(inParams.uri)
+    vlanName := pathInfo.Var("vlan")
```
Comment (venkatmahalingam) [inline src/translib/transformer/xfmr_fdb.go:36]: this should always be vlan-id, in which case, we get vlan-name?

File: src/translib/transformer/xfmr_fdb.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    return rmap, err
+}
+
+var YangToDb_entry_type_field_xfmr FieldXfmrYangToDb = func(inParams XfmrParams) (map[string]string, error) {
+    res_map := make(map[string]string)
+    var err error
+    /*if inParams.param == nil {
+        res_map[ENTRY_TYPE] = ""
+	return res_map, err
```
Comment (venkatmahalingam) [inline src/translib/transformer/xfmr_fdb.go]: If this function is not required, please remove the code.

File: src/translib/transformer/xfmr_fdb.go (hunk 6 above)
Comment (ViramShah91) [reply src/translib/transformer/xfmr_fdb.go:36]: While GET-all operation - transformer queries to DB and replies with VLAN-name that is present in Redis-db, in this case, the transformer will create uri with VLAN name and YangToDb will be called.

File: src/translib/transformer/xfmr_fdb.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    return res_map, err
+}
+
+var DbToYang_entry_type_field_xfmr FieldXfmrDbtoYang = func(inParams XfmrParams) (map[string]interface{}, error) {
+    var err error
+    result := make(map[string]interface{})
+    data := (*inParams.dbDataMap)[inParams.curDb]
+    fdbTableMap := data["FDB_TABLE"]
+    var entryTypeFinal = ""
+    for k, v := range fdbTableMap {
```
Comment (venkatmahalingam) [inline src/translib/transformer/xfmr_fdb.go]: do we need to walk the entire FDB table to know the key match? dont we have any better find function in golang?

File: src/translib/transformer/xfmr_fdb.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        if inParams.key == k {
+            entryType, ok := v.Field["type"]
+	    if ok {
+		entryTypeFinal = entryType
+	    }
+        }
+    }
+    oc_entrytype := findInMap(FDB_ENTRY_TYPE_MAP, entryTypeFinal)
```
Comment (venkatmahalingam) [inline src/translib/transformer/xfmr_fdb.go:91]: if entryTYpeFinal == "", do we still need construct the result and reply?

File: models/yang/annotations/sonic-fdb-annot.yang [hunk 10]
```diff
@@ -0,0 +1,17 @@
+module sonic-fdb-annot {
+
+    yang-version "";
+
+    namespace "http://openconfig.net/yang/annotation";
+    prefix "sfdb-annot";
+
+    import ietf-yang-types { prefix yang; }
```
Comment (venkatmahalingam) [inline models/yang/annotations/sonic-fdb-annot.yang]: Why do we need this sonic YANG annotation?

File: models/yang/sonic/sonic-fdb.yang [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        container sonic-fdb {
+
+                container FDB_TABLE {
+
+                        list FDB_TABLE_LIST {
+                                key "Vlan mac-address";
+
+                                leaf Vlan {
```
Comment (venkatmahalingam) [inline models/yang/sonic/sonic-fdb.yang]: Can you use "vlan" instead of "Vlan"?

File: src/translib/common_app.go [hunk 12]
```diff
@@ -235,9 +235,9 @@ func (app *CommonApp) translateCRUDCommon(d *db.DB, opcode int) ([]db.WatchKeys,
 	OrdTblList, err = transformer.GetOrdDBTblList(moduleNm)
 	if (err != nil) || (len(OrdTblList) == 0) {
 		log.Error("GetOrdDBTblList() failed")
-		return keys, err
+		//return keys, err
 	}
-
+        OrdTblList = []string {"FDB_TABLE"}
```
Comment (venkatmahalingam) [inline src/translib/common_app.go]: I dont think, this can be commited into project-arlo master, please ask infra team to commit the infra changes.

File: src/translib/transformer/xlate_from_db.go (hunk 5 above)
Comment (ViramShah91) [reply src/translib/transformer/xlate_from_db.go]: Removed this change. Infra team will merge this change in master branch.

File: src/translib/transformer/xfmr_fdb.go (hunk 7 above)
Comment (ViramShah91) [reply src/translib/transformer/xfmr_fdb.go]: Done

File: src/translib/transformer/xfmr_fdb.go (hunk 8 above)
Comment (ViramShah91) [reply src/translib/transformer/xfmr_fdb.go]: Done.

File: src/translib/transformer/xfmr_fdb.go (hunk 9 above)
Comment (ViramShah91) [reply src/translib/transformer/xfmr_fdb.go:91]: If a key is not found or entry-type final is ", we return without constructing reply.

File: models/yang/annotations/sonic-fdb-annot.yang (hunk 10 above)
Comment (ViramShah91) [reply models/yang/annotations/sonic-fdb-annot.yang]: The file is removed.

File: models/yang/sonic/sonic-fdb.yang (hunk 11 above)
Comment (ViramShah91) [reply models/yang/sonic/sonic-fdb.yang]: Done.

File: src/translib/common_app.go (hunk 12 above)
Comment (ViramShah91) [reply src/translib/common_app.go]: removed changes from my branch.. The infra team will address this issue.

File: models/yang/annotations/openconfig-network-instance-annot.yang [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+    deviation /oc-netinst:network-instances/oc-netinst:network-instance/oc-netinst:fdb/oc-netinst:mac-table/oc-netinst:entries/oc-netinst:entry/oc-netinst:state/oc-netinst:entry-type {
+      deviate add {
+            sonic-ext:field-transformer "entry_type_field_xfmr";
+      }
+    }
+
+    deviation /oc-netinst:network-instances/oc-netinst:network-instance/oc-netinst:fdb/oc-netinst:mac-table/oc-netinst:entries/oc-netinst:entry/oc-netinst:state/oc-netinst:vlan {
+      deviate add {
+            sonic-ext:field-name "Vlan";
```
Comment (venkatmahalingam) [inline models/yang/annotations/openconfig-network-instance-annot.yang]: field-name cant be "Vlan", just use "vlan", do the validation internally for user to give "Vlan"

File: models/yang/sonic/sonic-fdb.yang [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+                                        type yang:mac-address;
+                                }
+
+                                leaf port {
+                                        type string;
+                                }
+
+                                leaf type {
+                                        type string;
+                                }
```
Comment (venkatmahalingam) [inline models/yang/sonic/sonic-fdb.yang:55]: what's this type string? I dont infer anything from this attribute.

Comment (ViramShah91) [reply models/yang/sonic/sonic-fdb.yang:55]: Leaf "type" resides in Redis-db and it refers to "entry-type" as "static / dynamic". I have added description for all the yang attributes.

File: models/yang/annotations/openconfig-network-instance-annot.yang (hunk 13 above)
Comment (ViramShah91) [reply models/yang/annotations/openconfig-network-instance-annot.yang]: Done

**Review by venkatmahalingam — APPROVED** [review]

**Review by jeffyin-dell — APPROVED** [review]
